core/src/main/java/org/infinispan/factories/DataContainerFactory.java

```diff
@@ -34,6 +34,8 @@
 public class DataContainerFactory extends AbstractNamedCacheComponentFactory implements
          AutoInstantiableFactory {
 
+   private static final String SEGMENT_CONTAINER_FEATURE = "segmented-container";
```

Comment:
I mean, having a class called Features, for example, and adding methods like

 ```java
public boolean segmentContainer() {
     globalConfiguration.features().isAvailable(SEGMENT_CONTAINER_FEATURE)
}
```

Reply:
Thanks, I was debating about where to put it, but I figured for the first pass to leave it out until we came to a consensus. The reason I was a bit hesitant was I don't know how long each "feature" toggle will live for and didn't want to expose it yet, but I am fine with adding a method to https://github.com/infinispan/infinispan/blob/master/commons/src/main/java/org/infinispan/commons/util/Features.java#L48 similar to what you have in your example Katia. @tristantarrant WDYT?

Reply:
tbh, I don't think features deserve too many engineering :)
Also, does this disable segmented persistence ?

Reply:
@wburns @tristantarrant I'm ok with your choice

Reply:
This does not affect segmented persistence since that is configurable at the store level via the `segmented` attribute https://github.com/infinispan/infinispan/blob/master/core/src/main/java/org/infinispan/configuration/cache/StoreConfigurationChildBuilder.java#L102 

Segmented stores will always increase performance for stream operations, however it can slow down other operations such as put/get depending on the store implementation. Also many store implementations don't support segmentation. Due to these facts store segmentation is an actual exposed parameter.

Reply:
@tristantarrant ^

Reply:
Yes, we cannot "unexpose" the parameter, but we should "featurize" it

Reply:
@wburns are you going to add additional modifications then or should I just merge ?

Reply:
Hrmm, I personally am not a fan of "featurizing" the segment store part. What if a user sets a store to be "segmented" but then has the feature disabled? Do we then throw an error? Segmented stores are disabled by default and have to be explicitly enabled via the parameter.

Maybe you could clarify a bit more @tristantarrant ?

Reply:
That is exactly what I want for downstream: disable the ability to create a segmented store (throw an exception if the user attempts to enable that parameter)

Reply:
Alright, thanks for the clarification. What do you think we want the feature name to be then? Maybe `data-segmentation`? We can't really use data container anymore.

Reply:
+1 for `data-segmentation`, but separate `segmented-container` and `segmented-stores` would also work for me.

BTW, do we throw an exception if the user has a store with segmentation enabled and restarts with segmentation disabled (or the other way around)?

Reply:
There is currently no easy way to detect that between stores, so no there is not.

Reply:
Should we try to store some metadata like this in the store itself? I'd really like to have a clear exception when the configuration changes or an Infinispan upgrade changes the storage format in an incompatible way, but I'm not sure if this is best done with in-store metadata or with something in the global persistent state.

Reply:
Obviously not in 9.4. However the idea behind @ryanemerson 's work is to add this kind of metadata to aid in upgrade/migration scenarios and that can be also used to avoid incompatibility situations. Until that is in place, we'll just have to rely on warnings in the docs.